python/cudf/cudf/core/frame.py

```diff
@@ -42,129 +43,183 @@ def _concat(cls, objs, axis=0, ignore_index=False):
+            # Cast all columns to a common dtype, assign combined categories,
+            # and back-fill missing columns with all-null columns
+            for idx in col_idxs:
+                dtype = dtypes[idx]
+                for cols in list_of_columns:
+                    # If column not in this df, fill with an all-null column
+                    if idx >= len(cols) or cols[idx] is None:
+                        n = len(next(filter(lambda x: x is not None, cols)))
+                        cols[idx] = column_empty(n, dtype, masked=True)
                     else:
-                        non_null_columns[name].append(df[name])
-
-        for name, cols in non_null_columns.items():
-            # default to the first non-null dtype
-            dtypes[name] = cols[0].dtype
-            # If all the non-null dtypes are int/float, find a common dtype
-            if all(is_numerical_dtype(col.dtype) for col in cols):
-                dtypes[name] = np.find_common_type(
-                    [col.dtype for col in cols], []
-                )
-            # If all categorical dtypes, combine the categories
-            elif all(is_categorical_dtype(col.dtype) for col in cols):
-                # Combine and de-dupe the categories
-                categories[name] = (
-                    cudf.concat([col.cat.categories for col in cols])
-                    .to_series()
-                    .drop_duplicates()
-                    ._column
-                )
-                # Set the column dtype to the codes' dtype. The categories
-                # will be re-assigned at the end
-                dtypes[name] = min_scalar_type(len(categories[name]))
-            # Otherwise raise an error if columns have different dtypes
-            elif not all(is_dtype_equal(c.dtype, dtypes[name]) for c in cols):
-                raise ValueError("All columns must be the same type")
+                        # If column is categorical, rebase the codes with the
+                        # combined categories, and cast the new codes to the
+                        # min-scalar-sized dtype
+                        if idx in categories:
+                            cols[idx] = (
+                                cols[idx]
+                                .cat()._set_categories(
+                                    categories[idx], is_unique=True
+                                )
+                                .codes
+                            )
+                        cols[idx] = cols[idx].astype(dtype)
+
+        def reassign_categories(categories, cols, col_idxs):
+            for name, idx in zip(cols, col_idxs):
+                if idx in categories:
+                    cols[name] = build_categorical_column(
+                        categories=categories[idx],
+                        codes=cols[name],
+                        mask=cols[name].mask,
+                    )
```

Comment:
In case we have sliced inputs for whatever reason:
```suggestion
                    cols[name] = build_categorical_column(
                        categories=categories[idx],
                        codes=cols[name],
                        mask=cols[name].base_mask,
                        offset=cols[name].offset,
                        size=cols[name].size
                    )
```